src/BatchResult.php

```diff
@@ -0,0 +1,196 @@
+<?php
+
+namespace Http\Client;
+
+use Psr\Http\Message\RequestInterface;
+use Psr\Http\Message\ResponseInterface;
+
+/**
+ * Successful responses and exceptions returned from parallel request execution
+ *
+ * @author Márk Sági-Kazár <mark.sagikazar@gmail.com>
+ */
+final class BatchResult
+{
+    /**
+     * @var \SplObjectStorage
+     */
+    private $responses;
+
+    /**
+     * @var \SplObjectStorage
+     */
+    private $exceptions;
+
+    public function __construct()
+    {
+        $this->responses = new \SplObjectStorage();
+        $this->exceptions = new \SplObjectStorage();
+    }
+
+    /**
+     * Checks if a request is successful
+     *
+     * @param RequestInterface $request
+     *
+     * @return boolean
+     */
+    public function isSuccessful(RequestInterface $request)
+    {
+        return $this->responses->contains($request);
+    }
+
+    /**
+     * Checks if a request is failed
+     *
+     * @param RequestInterface $request
+     *
+     * @return boolean
+     */
+    public function isFailed(RequestInterface $request)
+    {
+        return $this->exceptions->contains($request);
+    }
+
+    /**
+     * Returns all successful responses
+     *
+     * @return ResponseInterface[]
+     */
+    public function getResponses()
+    {
+        $responses = [];
+
+        foreach ($this->responses as $request) {
+            $responses[] = $this->responses[$request];
+        }
+
+        return $responses;
+    }
+
+    /**
+     * Returns a response of a request or null if not found
+     *
+     * @param RequestInterface $request
+     *
+     * @return ResponseInterface|null
+     */
+    public function getResponseFor(RequestInterface $request)
+    {
+        if ($this->responses->contains($request)) {
+            return $this->responses[$request];
+        }
```

Comment:
hm. i would prefer the clean and logic code structure over insisting on
the immutability. if people want to do weird stuff, they always find
ways to do it.

we can do an `@internal` annotation on the add method and note you
should not use this outside a client/adapter.